File: backend/grant-service/api/src/main/kotlin/Gifts.kt
package dk.sdu.cloud.grant.api

import dk.sdu.cloud.AccessRight
import dk.sdu.cloud.CommonErrorMessage
import dk.sdu.cloud.FindByLongId
import dk.sdu.cloud.calls.CallDescriptionContainer
import dk.sdu.cloud.calls.*
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode

interface Gift {
    val title: String
    val description: String
    val resources: List<ResourceRequest>
    val resourcesOwnedBy: String
}

data class GiftWithId(
    val id: Long,
    override val resourcesOwnedBy: String,
    override val title: String,
    override val description: String,
    override val resources: List<ResourceRequest>
) : Gift

data class GiftWithCriteria(
    val id: Long,
    override val resourcesOwnedBy: String,
    override val title: String,
    override val description: String,
    override val resources: List<ResourceRequest>,
    val criteria: List<UserCriteria>
) : Gift {
    init {
        if (resources.isEmpty()) throw RPCException("resources cannot be empty", HttpStatusCode.BadRequest)
        if (criteria.isEmpty()) throw RPCException("resources cannot be empty", HttpStatusCode.BadRequest)
    }
}

data class ClaimGiftRequest(val giftId: Long)
typealias ClaimGiftResponse = Unit

typealias CreateGiftRequest = GiftWithCriteria
typealias CreateGiftResponse = FindByLongId

data class DeleteGiftRequest(val giftId: Long)
typealias DeleteGiftResponse = Unit

typealias AvailableGiftsRequest = Unit
data class AvailableGiftsResponse(val gifts: List<GiftWithId>)

typealias ListGiftsRequest = Unit
data class ListGiftsResponse(val gifts: List<GiftWithCriteria>)

object Gifts : CallDescriptionContainer("gifts") {
    const val baseContext = "/api/gifts"

    val claimGift = call<ClaimGiftRequest, ClaimGiftResponse, CommonErrorMessage>("claimGift") {
        auth {
            access = AccessRight.READ_WRITE
        }

        http {
            method = HttpMethod.Post

            path {
                using(baseContext)
                +"claim"
            }

            body { bindEntireRequestFromBody() }
        }
    }

    val availableGifts = call<AvailableGiftsRequest, AvailableGiftsResponse, CommonErrorMessage>("availableGifts") {
        auth {
            access = AccessRight.READ_WRITE
        }

        http {
            method = HttpMethod.Get

            path {
                using(baseContext)
                +"available"
            }
        }
    }

    val createGift = call<CreateGiftRequest, CreateGiftResponse, CommonErrorMessage>("createGift") {
        auth {
            access = AccessRight.READ_WRITE
        }

        http {
            method = HttpMethod.Post

            path {
                using(baseContext)
            }

            body { bindEntireRequestFromBody() }
        }
    }

    val deleteGift = call<DeleteGiftRequest, DeleteGiftResponse, CommonErrorMessage>("deleteGift") {
        auth {
            access = AccessRight.READ_WRITE
        }

        http {
            method = HttpMethod.Delete

            path {
                using(baseContext)
            }

            body { bindEntireRequestFromBody() }
        }
    }

    val listGifts = call<ListGiftsRequest, ListGiftsResponse, CommonErrorMessage>("listGifts") {
        auth {
            access = AccessRight.READ_WRITE
        }

        http {
            method = HttpMethod.Get

            path {
                using(baseContext)
            }
        }
    }
}

File: frontend-web/webclient/app/Accounting/Gifts.ts
import * as Grants from "@/Grants";
import {apiBrowse, apiCreate, apiDelete, apiUpdate} from "@/Authentication/DataHook";
import {FindByLongId, PageV2, PaginationRequestV2} from "@/UCloud";
import {removeTrailingSlash} from "@/UtilityFunctions";

const baseContext = "/api/gifts";

export interface Gift {
    title: string;
    description: string;
    resources: Grants.AllocationRequest[];
    resourcesOwnedBy: string;
}

export interface GiftWithCriteria extends Gift {
    id: number;
    criteria: Grants.UserCriteria[];
}

export function claim(request: { giftId: number }): APICallParameters {
    return apiUpdate(request, baseContext, "claim");
}

export function available(): APICallParameters<unknown, { gifts: FindByLongId[] }> {
    return {
        context: "",
        method: "GET",
        path: removeTrailingSlash(baseContext) + "/" + "available",
    };
}

export function create(request: GiftWithCriteria): APICallParameters<unknown, FindByLongId> {
    return apiCreate(request, baseContext);
}

export function remove(request: { giftId: number }): APICallParameters {
    return apiDelete(request, baseContext);
}

export function browse(request: PaginationRequestV2): APICallParameters<unknown, PageV2<GiftWithCriteria>> {
    return apiBrowse(request, baseContext);
}